rust_src/src/keymap.rs

```diff
@@ -62,6 +65,193 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
+    fun: map_keymap_function_t,
+    args: LispObject,
+    data: *mut c_void,
+) {
+    unsafe {
+        let map = safe_call1(Qkeymap_canonicalize, map);
+        map_keymap_internal(map, fun, args, data);
+    }
+}
+
+// Scopes are all messed up. Double check those
+
+/* Given OBJECT which was found in a slot in a keymap,
+trace indirect definitions to get the actual definition of that slot.
+An indirect definition is a list of the form
+(KEYMAP . INDEX), where KEYMAP is a keymap or a symbol defined as one
+and INDEX is the object to look up in KEYMAP to yield the definition.
+
+Also if OBJECT has a menu string as the first element,
+remove that.  Also remove a menu help string as second element.
+
+If AUTOLOAD, load autoloadable keymaps
+that are referred to with indirection.
+
+This can GC because menu_item_eval_property calls Feval.  */
+#[no_mangle]
+pub extern "C" fn _get_keyelt(object: LispObject, autoload: bool) -> LispObject {
+    loop {
+        if object.is_not_cons() {
+            break object;
+        }
+
+        // Convert the lisp object to a cons-cell in order to manipulate it
+        let consObject = object.as_cons().unwrap();
+
+        // Local mutable copy of object since we cannot modify the object
+        // variable passed to the function
+        let mut localObject = object;
+
+        if consObject.car().eq(Qmenu_item) {
+            if consObject.cdr().is_cons() {
+                // set object to (cdr (cdr object))
+                localObject = consObject.cdr().as_cons().unwrap().cdr();
+
+                if localObject.is_cons() {
+                    localObject = localObject.as_cons().unwrap().car();
+                }
+                let object_iter =
+                    localObject.iter_cars(LispConsEndChecks::on, LispConsCircularChecks::on);
+
+                // Iterate over every element in the list
+                for val in object_iter {
+                    if val.eq(QCfilter) && autoload {
+                        // Get the next object in the list
+                        let mut filter = val.as_cons().unwrap().cdr();
+
+                        unsafe {
+                            filter = list2(filter, list2(Qquote, localObject));
+                            localObject = menu_item_eval_property(filter);
+                        }
+                        break;
+                    }
+                }
+            } else {
+                // Return since object must be an invalid keymap
+                break localObject;
+            }
+        }
+        /* If the keymap contents looks like (STRING . DEFN), use DEFN.
+        Keymap alist elements like (CHAR MENUSTRING . DEFN)
+        will be used by HierarKey menus.  */
+        else if consObject.cdr().is_string() {
+            localObject = consObject.into();
+        } else {
+            break localObject;
+        }
+    }
+}
+
+// TODO: Fix this mess with all the unwrap but how?
+// TODO: Avoid using FCONS?
```

Comment:
Read my guidance earlier and suggestions above. Should make this code more manageable.